aptos-move/aptos-vm-types/src/storage/io_pricing.rs

```diff
@@ -166,26 +165,45 @@ impl StoragePricingV2 {
 
 // No storage curve. New gas parameter representation.
 #[derive(Debug, Clone)]
-pub struct StoragePricingV3 {
+pub struct IoPricingV3 {
     pub feature_version: u64,
-    pub free_write_bytes_quota: NumBytes,
+    pub legacy_free_write_bytes_quota: NumBytes,
 }
 
-impl StoragePricingV3 {
+impl IoPricingV3 {
+    fn read_size(&self, loaded: NumBytes) -> NumBytes {
+        if self.feature_version >= 12 {
+            // round up bytes to whole pages
+
+            const PAGE_SIZE: u64 = 4096;
+            let loaded_u64: u64 = loaded.into();
+            let r = loaded_u64 % PAGE_SIZE;
+            let rounded_up = loaded_u64 + if r == 0 { 0 } else { PAGE_SIZE - r };
+            NumBytes::new(rounded_up)
+        } else {
+            loaded
+        }
+    }
+
     fn calculate_read_gas(
         &self,
         loaded: NumBytes,
     ) -> impl GasExpression<VMGasParameters, Unit = InternalGasUnit> {
-        STORAGE_IO_PER_STATE_SLOT_READ * NumArgs::from(1) + STORAGE_IO_PER_STATE_BYTE_READ * loaded
+        STORAGE_IO_PER_STATE_SLOT_READ * NumArgs::from(1)
+            + STORAGE_IO_PER_STATE_BYTE_READ * self.read_size(loaded)
     }
 
     fn write_op_size(&self, key: &StateKey, value_size: u64) -> NumBytes {
         let key_size = NumBytes::new(key.size() as u64);
         let value_size = NumBytes::new(value_size);
+        let size = key_size + value_size;
 
-        (key_size + value_size)
-            .checked_sub(self.free_write_bytes_quota)
-            .unwrap_or(NumBytes::zero())
+        if self.feature_version >= 12 {
+            size.checked_sub(self.legacy_free_write_bytes_quota)
+                .unwrap_or(NumBytes::zero())
```

Comment:
shouldn't it be the opposite?

Reply:
:trollface: :trollface: :trollface: 
👍 👍 👍